tests/exporters/test_chirps.py

```diff
@@ -30,11 +34,56 @@ def test_checkpointing(self, mock_system, tmp_path, capsys):
         exporter = CHIRPSExporter(tmp_path)
 
-        africa_url = 'ftp://ftp.chg.ucsb.edu/pub/org/chg/products/CHIRPS-2.0/africa_pentad/tifs/'
-        global_url = 'ftp://ftp.chg.ucsb.edu/pub/org/chg/products/CHIRPS-2.0/global_pentad/netcdf/'
+        assert exporter.get_url('africa') == self.africa_url, \
+            f'Expected Africa URL to be {self.africa_url}, got {exporter.get_url("africa")}'
+
+        assert exporter.get_url('global') == self.global_url, \
+            f'Expected Africa URL to be {self.global_url}, got {exporter.get_url("global")}'
+
+    @patch('urllib.request.Request', autospec=True)
+    def test_get_filenames(self, request_patch, monkeypatch, tmp_path):
+        # First 1000 characters of the urllib response from the ftp for africa data,
+        # pulled on May 23 2019
+        request_patch.return_value = MagicMock()
+        expected_response = '-rw-r--r--    1 31094    31111     4172329 Feb 01  2015 ' \
+                            'chirps-v2.0.1981.01.1.tif.gz\r\n-rw-r--r--    1 31094    ' \
+                            '31111     4133059 Feb 01  2015 chirps-v2.0.1981.01.2.tif.' \
+                            'gz\r\n-rw-r--r--    1 31094    31111     3948001 Feb 01  20' \
+                            '15 chirps-v2.0.1981.01.3.tif.gz\r\n-rw-r--r--    1 31094    ' \
+                            '31111     3942431 Feb 01  2015 chirps-v2.0.1981.01.4.tif.g' \
+                            'z\r\n-rw-r--r--    1 31094    31111     4176578 Feb 01  201' \
+                            '5 chirps-v2.0.1981.01.5.tif.gz\r\n-rw-r--r--    1 31094    ' \
+                            '31111     4333592 Feb 01  2015 chirps-v2.0.1981.01.6.tif.' \
+                            'gz\r\n-rw-r--r--    1 31094    31111     4129989 Feb 02  2015' \
+                            ' chirps-v2.0.1981.02.1.tif.gz\r\n-rw-r--r--    1 31094    3' \
+                            '1111     4067482 Feb 02  2015 chirps-v2.0.1981.02.2.tif.' \
+                            'gz\r\n-rw-r--r--    1 31094    31111     3927390 Feb 02  2015' \
+                            ' chirps-v2.0.1981.02.3.tif.gz\r\n-rw-r--r--    1 31094    311' \
+                            '11     4208427 Feb 02  2015 chirps-v2.0.1981.02.4.tif.gz\r\n-' \
+                            'rw-r--r--    1 31094    31111     4301096 Feb 02  2015 chirps-v' \
+                            '2.0.1981.02.5.tif.gz\r\n'
+
+        expected_filenames = ['chirps-v2.0.1981.01.1.tif.gz', 'chirps-v2.0.1981.01.2.tif.gz',
+                              'chirps-v2.0.1981.01.3.tif.gz', 'chirps-v2.0.1981.01.4.tif.gz',
+                              'chirps-v2.0.1981.01.5.tif.gz', 'chirps-v2.0.1981.01.6.tif.gz',
+                              'chirps-v2.0.1981.02.1.tif.gz', 'chirps-v2.0.1981.02.2.tif.gz',
+                              'chirps-v2.0.1981.02.3.tif.gz', 'chirps-v2.0.1981.02.4.tif.gz',
+                              'chirps-v2.0.1981.02.5.tif.gz']
+
+        def mockreturn(request):
```

Comment:
@gabrieltseng this is far out mate! Can you explain to me how it works?

Reply:
The functionality I was testing for was that given a string (the response body of the africa url), the code would correctly find the correct filenames.

To do this, I used [`monkeypatch`](https://stackoverflow.com/questions/5626193/what-is-monkey-patching), which replaces what a function normally does with whatever you want it to do.

I wanted this line: `response = urllib.request.urlopen(req)` to return a class which could be `.read()` to get the first ~1000 characters of the string, instead of opening a connection with the url, so I replaced its original functionality with `mockreturn`

Reply:
niiiice! does it matter what you call the class:
```
def mockreturn(request):
    class OpenURL:
        def read(self):
            return expected_response
    open_url = OpenURL()
    return open_url
```

Also does the following therefore mean that any calls to `urllib.request.urlopen` are replaced with mockreturn?

```
monkeypatch.setattr(urllib.request, 'urlopen', mockreturn)
exporter = CHIRPSExporter(tmp_path)
filenames = exporter.get_chirps_filenames(region='africa')
```


```

Reply:
The class can be called anything, and yes any calls to `urllib.request.urlopen` are replaced with mockreturn.